WordPress/Classes/Models/Blog+SelfHosted.swift

```diff
@@ -0,0 +1,132 @@
+        case invalidCredentialsUrl
+        case invalidXmlRpcEndpoint
+    }
+
+    static func createRestApiBlog(
+        with details: SelfHostedLoginDetails,
+        in contextManager: ContextManager,
+        using keychainImplementation: KeychainAccessible = KeychainUtils()
+    ) async throws -> String {
+        try await contextManager.performAndSave { context in
+            let blog = Blog.createBlankBlog(in: context)
+            blog.setUrl(details.url)
+            blog.username = details.username
+            try blog.setPassword(to: details.password, using: keychainImplementation)
+            blog.setXMLRPCEndpoint(to: details.derivedXMLRPCRoot)
+            blog.setSiteIdentifier(details.derivedSiteId)
+
+            return details.derivedSiteId
+        }
+    }
+
+    static func lookupRestApiBlog(with id: SiteIdentifier, in context: NSManagedObjectContext) throws -> Blog? {
+        try BlogQuery().apiKey(is: id).blog(in: context)
+    }
+
+    static func hasRestApiBlog(with id: SiteIdentifier, in context: NSManagedObjectContext) throws -> Bool {
+        BlogQuery().apiKey(is: id).count(in: context) != 0
+    }
+
+    // MARK: Type-safe wrappers
+    // The underlying `Blog` object has lots of field nullability that doesn't provide guarantees about
+    // which fields are present. These wrappers will `throw` if the `Blog` is invalid, allowing any dependent
+    // code can be much simpler.
+
+    /// An alias for `getPassword` to make it obvious how to retrieve Application Tokens
+    ///
+    func getApplicationToken(using keychainImplementation: KeychainAccessible = KeychainUtils()) throws -> String {
+        try getPassword(using: keychainImplementation)
+    }
+
+    /// An alias for `setPassword` to make it obvious how to store Application Tokens
+    ///
+    func setApplicationToken(
+        _ newValue: String,
+        using keychainImplementation: KeychainAccessible = KeychainUtils()
+    ) throws {
+        try setPassword(to: newValue, using: keychainImplementation)
+    }
+
+    /// A null-safe wrapper for `Blog.username`
+    func getUsername() throws -> String {
+        guard let username = self.username else {
+            throw BlogCredentialsError.blogUsernameMissing
+        }
+
+        return username
+    }
+
+    /// A null-safe replacement for `Blog.password(get)`
+    func getPassword(using keychainImplementation: KeychainAccessible = KeychainUtils()) throws -> String {
+        try keychainImplementation.getPassword(for: self.getUsername(), serviceName: self.getUrlString())
+    }
+
+    /// A null-safe replacement for `Blog.password(set)`
+    func setPassword(to newValue: String, using keychainImplementation: KeychainAccessible = KeychainUtils()) throws {
+        try keychainImplementation.setPassword(for: self.getUsername(), to: newValue, serviceName: self.getUrlString())
+    }
+
+    /// A null-and-type-safe replacement for `Blog.url(get)`
+    func getUrl() throws -> URL {
```

Comment:
- (nit) should be spelled `URL` instead of `Url`

Reply:
(General comment)

If the goal is to use these only to create `SelfHostedLoginDetails `, I would suggest making these APIs private and/or inlining in `SelfHostedLoginDetails`. The `Blog`, which is supposed to be a simple database object, already has hundreds of APIs, and it's really hard to know what you should or shouldn't use. I would be cautious about extending it even further - it would be best to do the opposite.

The errors also appear to be a bit too granular. If we are no planning to handle these individual cases anywhere, it's best to simply model these fields as optionals and have a generic way to unwrap them like `XCTUnwrap` and provide a generic error message. Adding a separate method and a separate error case anywhere you have an optional will not scale.

Example:

```swift
static func from(blog: Blog) throws -> SelfHostedLoginDetails {
         return try SelfHostedLoginDetails(
             url: expect(blog.url, "url_missing_or_invalid"),

// or for conversion to URL
             url: expect(blog.url.flatMap(URL.init), "url_missing_or_invalid"),
```

Reply:
I think it's also a good use-case for a macro:

```swift
#expect(blog.url) // generates error message automatically: "missing `blog.url`" with stringily (should be possible?)
```

Reply:
> should be spelled URL instead of Url

In this case it's just `camelCase('get', 'url')` because `url` is the name of the underlying property. 

In general I'd like to standardize it (there's ~10x the `URL` as `Url` but it'd take _forever_ to fix that)

Reply:
> Blog, which is supposed to be a simple database object, already has hundreds of APIs, and it's really hard to know what you should or shouldn't use. I would be cautious about extending it even further - it would be best to do the opposite.

I agree, but propagating the self-hosted complexity further throughout the app doesn't help us. Enforcing consistency makes it impossible to get bad data into Core Data, which will matter a lot when we go to replace it later.

> The errors also appear to be a bit too granular. If we are no planning to handle these individual cases anywhere, it's best to simply model these fields as optionals and have a generic way to unwrap them like XCTUnwrap with generic string-based errors. Adding a separate method and a separate error case anywhere you have an optional will not scale.

In this case, none of these error types should ever be emitted (that would be a result of programmer error) but it'll help with migration and uniquely identify them in Sentry. Once we have a proper data layer in place, we can drop them entirely (because the data model will enforce correctness at that point).

Reply:
> Adding a separate method and a separate error case anywhere you have an optional will not scale.

There should only be two destinations for this:

1. Sentry (in which case we just care about the type and the stack trace) and
2. The UI layer, where we can give better error info

So I don't think there's a ton of scaling here needed?